### ISSUE TITLE
Isuse with nchan and Android clients

### ISSUE THREAD

**Issue by SimonFair** [slact/nchan#636]
Currently I have noticed that the number of stored messages increases until memory is exhausted, issue is from Android clients and then selecting another tab in the client browser. Tried to set expiry for messages but then dont seem to get purged.

	#
	# Listen on local socket for nchan publishers
	#
	server {
	    listen unix:/var/run/nginx.socket default_server;
	    location ~ /pub/(.*)$ {
	        nchan_publisher;
	        nchan_channel_id "$1";
	        nchan_message_buffer_length $arg_buffer_length;
	        nchan_message_timeout 15s;
	    }
	    location ~ /nchan_stub_status$ {
	        nchan_stub_status;
	    }
	}

Every 1.0s: curl --unix-socket /var/run/nginx.socket http://localhost/nchan_stub_status                                                                                      computenode: Sun Mar 20 13:48:47 2022

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
   0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0 100   384  100   384    0     0  3289k      0 --:--:-- --:--:-- --:--:--  375k
total published messages: 42200
stored messages: 1879
shared memory used: 14908K
shared memory limit: 131072K
channels: 27
subscribers: 9
redis pending commands: 0
redis connected servers: 0
total interprocess alerts received: 0
interprocess alerts in transit: 0
interprocess queued alerts: 0
total interprocess send delay: 0
total interprocess receive delay: 0
nchan version: 1.2.15

**Comment by SimonFair** [comment]
Adding suggestions of ideas to try?

**Comment by slact** [comment]
How much shared memory does Nchan have? What is the message publishing rate and average size?

**Comment by SimonFair** [comment]
nchan has 131072K allocated.  

When it is running as expected outstanding messages are low

total published messages: 323326
stored messages: 8
shared memory used: 72K
shared memory limit: 131072K

Issue comes when an android device or similar is used on the page and the device goes stops processing the page, the messages start to build up to the point it exceeds the available memory. The subscriber is still connected.

I have tried add buffer length and timeout to the server but does not seem to make a difference.

      #
        # Listen on local socket for nchan publishers
        #
        server {
            listen unix:/var/run/nginx.socket default_server;
            location ~ /pub/(.*)$ {
                nchan_publisher;
                nchan_channel_id "$1";
                nchan_message_buffer_length 1;
                nchan_message_timeout 1s;
            }
            location ~ /nchan_stub_status$ {
                nchan_stub_status;
            }
        }

Message sizes are smallish around 4k

total published messages: 333099
stored messages: 462
shared memory used: 632K
shared memory limit: 131072K
channels: 15
subscribers: 8
redis pending commands: 0
redis connected servers: 0
redis unhealthy upstreams: 0
total redis commands sent: 0
total interprocess alerts received: 0
interprocess alerts in transit: 0
interprocess queued alerts: 0
total interprocess send delay: 0
total interprocess receive delay: 0
nchan version: 1.3.6

**Comment by slact** [comment]
What subscriber type do these devices use?

**Comment by slact** [comment]
Yes, thanks. What about the subscriber location block in your config?

**Comment by SimonFair** [comment]
#
    # Listen on local socket for nchan publishers
    #
    server {
        listen unix:/var/run/nginx.socket default_server;
        location ~ /pub/(.*)$ {
            nchan_publisher;
            nchan_channel_id "$1";
            nchan_message_buffer_length 1;
            nchan_message_timeout 1s;
        }
        location ~ /nchan_stub_status$ {
            nchan_stub_status;
        }
    }

**Comment by slact** [comment]
That's the _publisher_ location. I'm interested in the _subscriber_ location, the one containing `nchan_subscriber` or `nchan_pubsub`.

**Comment by SimonFair** [comment]
#
        # nchan subscriber endpoint
        #
        location ~ /sub/(.*)$ {
            nchan_subscriber;
            # nchan_authorize_request <url here>
            nchan_channel_id "$1";
            nchan_channel_id_split_delimiter ",";
        }
        location /nchan_stub_status {
            nchan_stub_status;
        }

**Comment by SimonFair** [comment]
@slact any suggestions on what to try?

**Comment by slact** [comment]
If your traffic permits it, please run Nchan with 1 worker process and see if that makes a difference.

**Comment by SimonFair** [comment]
> If your traffic permits it, please run Nchan with 1 worker process and see if that makes a difference.

grep worker_processes /etc/nginx/nginx.conf
worker_processes  1;
